src/compute/src/sink/copy_to_s3_oneshot.rs

```diff
@@ -66,73 +97,24 @@ where
-    sinked_collection: Collection<G, Row, Diff>,
-    err_collection: Collection<G, DataflowError, Diff>,
-    sink_id: GlobalId,
-    up_to: Antichain<G::Timestamp>,
-    response_protocol_handle: Rc<RefCell<Option<ResponseProtocol>>>,
-) where
-    G: Scope<Timestamp = Timestamp>,
-{
-    let name = format!("s3-oneshot-{}", sink_id);
-    let mut op = OperatorBuilder::new(name, sinked_collection.scope());
-    let mut ok_input = op.new_input(&sinked_collection.inner, Pipeline);
-    let mut err_input = op.new_input(&err_collection.inner, Pipeline);
-
-    op.build(|_cap| {
-        let mut row_count: u64 = 0;
-        let mut finished = false;
-        let mut ok_buf = Default::default();
-
-        move |frontiers| {
-            if finished {
-                // Drain the inputs, to avoid the operator being constantly rescheduled
-                ok_input.for_each(|_, _| {});
-                err_input.for_each(|_, _| {});
-                return;
-            }
-
-            let mut frontier = Antichain::new();
-            for input_frontier in frontiers {
-                frontier.extend(input_frontier.frontier().iter().copied());
-            }
-
-            ok_input.for_each(|_, data| {
-                data.swap(&mut ok_buf);
-                let count: u64 = ok_buf.drain(..).len().try_into().expect("usize into u64");
-                // TODO (mouli): write these to s3 in subsequent PRs
-                row_count = row_count + count;
-            });
-            // TODO (mouli): handle errors
-            if PartialOrder::less_equal(&up_to, &frontier) {
-                finished = true;
-                // We are done, send the final count.
-                if let Some(response_protocol) = response_protocol_handle.borrow_mut().deref_mut() {
-                    response_protocol.send_final_count(row_count);
-                }
-            }
-        }
-    });
-}
-
 /// A type that guides the transmission of number of rows back to the coordinator.
 struct ResponseProtocol {
     pub sink_id: GlobalId,
     pub response_buffer: Option<Rc<RefCell<Vec<(GlobalId, CopyToResponse)>>>>,
 }
 
 impl ResponseProtocol {
-    fn send_final_count(&mut self, rows: u64) {
-        let buffer = self.response_buffer.as_mut().expect("Copy response buffer");
-
-        buffer
-            .borrow_mut()
-            .push((self.sink_id, CopyToResponse::RowCount(rows)));
-
+    // This method should only be called once.
+    fn send(&mut self, count: Result<u64, String>) {
         // The dataflow's input has been exhausted, clear the channel,
         // to avoid sending `CopyToResponse::Dropped`.
-        self.response_buffer = None;
+        if let Some(buffer) = self.response_buffer.take() {
+            let response = match count {
+                Ok(count) => CopyToResponse::RowCount(count),
+                Err(error) => CopyToResponse::Error(error),
+            };
+            buffer.borrow_mut().push((self.sink_id, response));
+        }
```

Comment:
Before we panicked when the `response_buffer` was already taken, now we silently ignore it. I think we still want to panic, or at least soft-panic, since this method being called twice is a bug.